cpp/include/patterns/copy_v_transform_reduce_nbr.cuh

```diff
@@ -0,0 +1,293 @@
+
+    grid_1d_thread_t update_grid(
+      matrix_partition.get_major_last() - matrix_partition.get_major_first(),
+      detail::copy_v_transform_reduce_nbr_for_all_low_out_degree_block_size,
+      get_max_num_blocks_1D());
+
+    if (!GraphViewType::is_adj_matrix_transposed) {
+      thrust::fill(rmm::exec_policy(handle.get_stream())->on(handle.get_stream()),
+                   vertex_value_output_first,
+                   vertex_value_output_first + graph_view.get_number_of_local_vertices(),
+                   init);
+    }
+
+    assert(graph_view.get_number_of_local_vertices() ==
+           graph_view.get_number_of_adj_matrix_local_rows());
+    assert(graph_view.get_number_of_local_vertices() ==
+           graph_view.get_number_of_adj_matrix_local_cols());
+    detail::for_all_major_for_all_nbr_low_out_degree<GraphViewType::is_adj_matrix_transposed>
+      <<<update_grid.num_blocks, update_grid.block_size, 0, handle.get_stream()>>>(
+        matrix_partition,
+        adj_matrix_row_value_input_first,
+        adj_matrix_col_value_input_first,
+        vertex_value_output_first,
+        e_op,
+        init);
+  }
+}
+
+/**
+ * @brief Iterate over the outgoing edges to update vertex properties.
+ *
+ * This function is inspired by thrust::transfrom_reduce() (iteration over the outgoing edges part)
+ * and thrust::copy() (update vertex properties part, take transform_reduce output as copy input).
+ *
+ * @tparam GraphViewType Type of the passed non-owning graph object.
+ * @tparam AdjMatrixRowValueInputIterator Type of the iterator for graph adjacency matrix row
+ * input properties.
+ * @tparam AdjMatrixColValueInputIterator Type of the iterator for graph adjacency matrix column
+ * input properties.
+ * @tparam EdgeOp Type of the binary (or ternary) edge operator.
+ * @tparam T Type of the initial value for reduction over the outgoing edges.
+ * @tparam VertexValueOutputIterator Type of the iterator for vertex output property variables.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Non-owning graph object.
+ * @param adj_matrix_row_value_input_first Iterator pointing to the adjacency matrix row input
+ * properties for the first (inclusive) row (assigned to this process in multi-GPU).
+ * `adj_matrix_row_value_input_last` (exclusive) is deduced as @p adj_matrix_row_value_input_first +
+ * @p graph_view.get_number_of_adj_matrix_local_rows().
+ * @param adj_matrix_col_value_input_first Iterator pointing to the adjacency matrix column input
+ * properties for the first (inclusive) column (assigned to this process in multi-GPU).
+ * `adj_matrix_col_value_output_last` (exclusive) is deduced as @p adj_matrix_col_value_output_first
+ * + @p graph_view.get_number_of_adj_matrix_local_cols().
+ * @param e_op Binary (or ternary) operator takes *(@p adj_matrix_row_value_input_first + i), *(@p
```

Comment:
Seems to really need 4 or 5 parameters: src, dest, optional weight, src_val, dst_val

Reply:
Yes, the code is fixed when updating this PR to work with the new graph class, I will fix the comment.

Reply:
Fixed.